[WebGPU] webgpu:api,operation,command_buffer,queries,* does not pass

mwyrzykowski opened the pull request:
#### 14d4ec3b58ca7115c18575f8e1c04091d9f37105
<pre>
[WebGPU] webgpu:api,operation,command_buffer,queries,* does not pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=264924">https://bugs.webkit.org/show_bug.cgi?id=264924</a>
&lt;radar://118492272&gt;

Reviewed by Tadeu Zagallo.

The spec requires occlusion queries to be zero if no geometry
is rendered between beginOcclusionQuery / endOcclusionQuery.

Also correct issues preventing sampleMask from working with RenderBundles
which were preventing occlusion queries using RenderBundles from passing
when changing the sample mask values.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::ensureBlitCommandEncoder):
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::endCurrentICB):
(WebGPU::RenderBundleEncoder::icbNeedsToBeSplit):
(WebGPU::RenderBundleEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
(WebGPU::RenderPassEncoder::create):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::beginOcclusionQuery):
(WebGPU::setViewportMinMaxDepthIntoBuffer):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::endPass):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::sampleMask const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::convertPipelineLayout):
* LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/queries/occlusionQuery-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/270934@main">https://commits.webkit.org/270934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cea8c8565cf3f32d303999d859a536bd82b40d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26835 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5451 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24525 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7288 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2842 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24412 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27097 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4264 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23030 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3743 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3794 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24027 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29677 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24469 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30026 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3816 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2014 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27933 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5274 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6453 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4283 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4184 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -74,13 +75,14 @@
 {
     static_assert(sizeof(fragmentDynamicOffsets[0]) == sizeof(minDepth), "expect dynamic offsets container to have matching size to depth values");
-    static_assert(startIndexForFragmentDynamicOffsets == 2, "code path assumes value is 2");
+    static_assert(startIndexForFragmentDynamicOffsets == 3, "code path assumes value is 3");
```
This should probably be > 1, similar to how you wrote below, as the actual code didn't have to change

tadeuzagallo commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -653,8 +653,11 @@ static WGPUTextureViewDimension convertViewDimension(WGSL::TextureViewDimension
     mtlRenderPipelineDescriptor.alphaToCoverageEnabled = descriptor.multisample.alphaToCoverageEnabled;
     RELEASE_ASSERT([mtlRenderPipelineDescriptor respondsToSelector:@selector(setSampleMask:)]);
-    if (auto mask = descriptor.multisample.mask; mask != 0xFFFFFFFF)
+    uint32_t sampleMask = UINT32_MAX;
```
Should this use `RenderBundleEncoder::defaultSampleMask` instead?

tadeuzagallo commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -369,7 +369,7 @@ static auto wgslViewDimension(WGPUTextureViewDimension viewDimension)
             wgslEntry.fragmentArgumentBufferSizeIndex = entry.value.bufferSizeArgumentBufferIndices[WebGPU::ShaderStage::Fragment];
             if (entry.value.fragmentDynamicOffset)
-                wgslEntry.fragmentBufferDynamicOffset = *entry.value.fragmentDynamicOffset + 2;
+                wgslEntry.fragmentBufferDynamicOffset = *entry.value.fragmentDynamicOffset + 3;
```
I might be nicer to use `startIndexForFragmentDynamicOffsets` here

mwyrzykowski committed:
[WebGPU] webgpu:api,operation,command_buffer,queries,* does not pass
https://bugs.webkit.org/show_bug.cgi?id=264924
<radar://118492272>

Reviewed by Tadeu Zagallo.

The spec requires occlusion queries to be zero if no geometry
is rendered between beginOcclusionQuery / endOcclusionQuery.

Also correct issues preventing sampleMask from working with RenderBundles
which were preventing occlusion queries using RenderBundles from passing
when changing the sample mask values.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::ensureBlitCommandEncoder):
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::endCurrentICB):
(WebGPU::RenderBundleEncoder::icbNeedsToBeSplit):
(WebGPU::RenderBundleEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
(WebGPU::RenderPassEncoder::create):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::beginOcclusionQuery):
(WebGPU::setViewportMinMaxDepthIntoBuffer):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::endPass):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::sampleMask const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::convertPipelineLayout):
* LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/queries/occlusionQuery-expected.txt: Added.

Canonical link: https://commits.webkit.org/270934@main
